src/ess/imaging/normalize.py

```diff
@@ -0,0 +1,235 @@
+
+from typing import NewType
+
+import scipp as sc
+
+from .io import (
+    TIME_COORD_NAME,
+    DarkCurrentImageStacks,
+    OpenBeamImageStacks,
+    SampleImageStacks,
+)
+
+AverageBackgroundPixelCounts = NewType("AverageBackgroundPixelCounts", sc.Variable)
+"""~math:`D0 = mean(background)`."""
+AverageSamplePixelCounts = NewType("AverageSamplePixelCounts", sc.Variable)
+"""~math:`D = mean(sample)`."""
+ScaleFactor = NewType("ScaleFactor", sc.Variable)
+"""~math:`ScaleFactor = AverageBackgroundPixelCounts / AverageSamplePixelCounts`."""
+
+OpenBeamImage = NewType("OpenBeamImage", sc.DataArray)
+"""Open beam image. ~math:`average(open_beam)`."""
+DarkCurrentImage = NewType("DarkCurrentImage", sc.DataArray)
+"""Dark current image ~math:`average(dark_current)`."""
+AveragedNonSampleImages = NewType("AveragedNonSampleImages", sc.DataArray)
+"""Averaged open beam and dark current images."""
+BackgroundImage = NewType("BackgroundImage", sc.DataArray)
+"""Background image stack. ~math:`background = open_beam - dark_current`."""
+CleansedSampleImages = NewType("CleansedSampleImages", sc.DataArray)
+"""Sample image stack - dark current."""
+NormalizedSampleImages = NewType("NormalizedSampleImages", sc.DataArray)
+"""Normalized sample image stack.
+
+~math:`normalized_sample = sample / background * DFactor`.
+"""
+GroupedSampleImages = NewType("GroupedSampleImages", sc.DataArray)
+"""Grouped sample image stack by rotation angle."""
+
+
+def _mean_all_dims(data: sc.Variable) -> sc.Variable:
+    """Calculate the mean of all dimensions one by one to avoid overflow."""
+    if data.shape == ():  # scalar
+        return data
+    return _mean_all_dims(data.mean(dim=data.dims[0]))
+
+
+def average_open_beam_images(open_beam: OpenBeamImageStacks) -> OpenBeamImage:
+    """Average the open beam image stack.
+
+    .. math::
+
+        OpenBeam = mean(open_beam, 'time')
+
+    """
+    return OpenBeamImage(sc.mean(open_beam, dim=TIME_COORD_NAME))
+
+
+def average_dark_current_images(
+    dark_current: DarkCurrentImageStacks,
+) -> DarkCurrentImage:
+    """Average the dark current image stack.
+
+    .. math::
+
+        DarkCurrent = mean(dark_current, 'time')
+
+    """
+    return DarkCurrentImage(sc.mean(dark_current, dim=TIME_COORD_NAME))
+
+
+BackgroundPixelThreshold = NewType("BackgroundPixelThreshold", sc.Variable)
+"""Threshold for the background pixel values."""
+
+DEFAULT_BACKGROUND_THRESHOLD = BackgroundPixelThreshold(sc.scalar(1.0, unit="counts"))
+
+
+def calculate_white_beam_background(
+    open_beam: OpenBeamImage,
+    dark_current: DarkCurrentImage,
+    background_threshold: BackgroundPixelThreshold = DEFAULT_BACKGROUND_THRESHOLD,
+) -> BackgroundImage:
+    """Calculate the background image stack.
+
+    We average the open beam and dark current image stack
+    to create the single background image.
+
+    .. math::
+
+        Background = mean(OpenBeam, 'time') - mean(DarkCurrent, 'time')
+
```

Comment:
Can you also add (in words or in the math) the fact that you are applying a threshold after the operation?